File: 211/p211.py
from math import sqrt

def sigma_x_sieve(m, x):
	buf = [1]*m
	for p in range(2, m):
		if buf[p] != 1:
			continue
		px = p**x
		buf[p] = px + 1
		for n in range(2*p, m, p):
			pxa = px
			d = n
			while d%p == 0:
				pxa *= px
				d /= p
			buf[n] *= (pxa - 1)//(px - 1)
	return buf

def is_square(n):
	r = int(sqrt(n))
	return r**2 == n

#M = 64000000
M = 640000

s = sum(n for (n, s) in enumerate(sigma_x_sieve(M, 2)[1:], 1) if is_square(s))
print(s)

File: 211/p211.hs
import Data.Array

sigma_x_sieve x buf p = if p > (snd $ bounds buf)
                        then buf
                        else if buf!p /= 1
                             then sigma_x_sieve x buf (p + 1)
                             else sigma_x_sieve x (buf // [(m, buf!m * sigma_x_factor x m p) | m <- [p,2*p..snd $ bounds buf]]) (p + 1)

sigma_x_factor x m p = div (p^(x*a + x) - 1) (p^x - 1)
                       where a = max_pow m p

max_pow m p = if m `mod` p == 0
              then 1 + max_pow (m `div` p) p
              else 0

is_square n = n == (truncate $ sqrt $ fromIntegral n)^2

sum_square_sigma_2s m = sum $ filter (is_square.(sigma_xs!)) [1..m]
                        where sigma_xs = sigma_x_sieve 2 (array (1, m) [(i, 1) | i <- [1..m]]) 2

main = do
    print $ sum_square_sigma_2s 640000

